service/history/replication/stream_sender_flow_controller.go

```diff
@@ -0,0 +1,135 @@
+	replicationpb "go.temporal.io/server/api/replication/v1"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+	"go.temporal.io/server/common/quotas"
+	"go.temporal.io/server/service/history/configs"
+)
+
+type (
+	flowControlState struct {
+		mu          sync.Mutex
+		cond        *sync.Cond
+		waiters     int
+		resume      bool
+		rateLimiter quotas.RateLimiter // todo: consider using a shared rate limiter across shard for better resource allocation
+	}
+	SenderFlowController interface {
+		// Wait will block go routine until the sender is allowed to send a task
+		Wait(priority enums.TaskPriority)
+		RefreshReceiverFlowControlInfo(syncState *replicationpb.SyncReplicationState)
+	}
+	SenderFlowControllerImpl struct {
+		flowControlStates  map[enums.TaskPriority]*flowControlState
+		defaultRateLimiter quotas.RateLimiter
+		logger             log.Logger
+	}
+)
+
+func NewSenderFlowController(config *configs.Config, logger log.Logger) *SenderFlowControllerImpl {
+	flowControlStates := make(map[enums.TaskPriority]*flowControlState)
+	highPriorityState := &flowControlState{
+		resume: true,
+	}
+	highPriorityState.cond = sync.NewCond(&highPriorityState.mu)
+	highPriorityState.rateLimiter = quotas.NewDynamicRateLimiter(quotas.NewDefaultOutgoingRateLimiter(func() float64 {
+		return float64(config.ReplicationStreamSenderHighPriorityMaxQPS())
+	}), 1*time.Minute)
+
+	lowPriorityState := &flowControlState{
+		resume: true,
+	}
+	lowPriorityState.cond = sync.NewCond(&lowPriorityState.mu)
+	lowPriorityState.rateLimiter = quotas.NewDynamicRateLimiter(quotas.NewDefaultOutgoingRateLimiter(func() float64 {
+		return float64(config.ReplicationStreamSenderLowPriorityMaxQPS())
+	}), 1*time.Minute)
+	flowControlStates[enums.TASK_PRIORITY_HIGH] = highPriorityState
+	flowControlStates[enums.TASK_PRIORITY_LOW] = lowPriorityState
+	return &SenderFlowControllerImpl{
+		flowControlStates:  flowControlStates,
+		defaultRateLimiter: quotas.NewRateLimiter(float64(config.ReplicationStreamSenderHighPriorityMaxQPS()), config.ReplicationStreamSenderHighPriorityMaxQPS()),
+		logger:             logger,
+	}
+}
+
+func (s *SenderFlowControllerImpl) RefreshReceiverFlowControlInfo(syncState *replicationpb.SyncReplicationState) {
+	if syncState.GetHighPriorityState() != nil {
+		s.setState(s.flowControlStates[enums.TASK_PRIORITY_HIGH], syncState.GetHighPriorityState().GetFlowControlCommand())
+	}
+	if syncState.GetLowPriorityState() != nil {
+		s.setState(s.flowControlStates[enums.TASK_PRIORITY_LOW], syncState.GetLowPriorityState().GetFlowControlCommand())
+	}
+}
+
+func (s *SenderFlowControllerImpl) setState(state *flowControlState, flowControlCommand enums.ReplicationFlowControlCommand) {
+	switch flowControlCommand {
+	case enums.REPLICATION_FLOW_CONTROL_COMMAND_RESUME, enums.REPLICATION_FLOW_CONTROL_COMMAND_UNSPECIFIED:
+		state.mu.Lock()
+		defer state.mu.Unlock()
+		state.resume = true
+		if state.waiters > 0 {
+			state.cond.Signal()
```

Comment:
From the code looks like there will be only one waiter for each priority. However this is not explicit given the mismatch of `waiters` (int) and `Signal` here (should be `Broadcast`?).

Reply:
Right, let's use Broadcast here to avoid future regression if multiple thread waiting on this `cond`.